lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -0,0 +1,132 @@
+        ArrayKeys: { className: "Array", methodName: "keys", argumentsCount: 0, forceInline: true /*optional*/ },
+        ArrayEntries: { className: "Array", methodName: "entries", argumentsCount: 0, forceInline: true /*optional*/ }
+    };
+
+    var setPrototype = platform.builtInSetPrototype;
+    var _objectDefineProperty = platform.builtInJavascriptObjectEntryDefineProperty;
+
+    // Object's getter and setter can get overriden on the prototype, in that case while setting the value attributes, we will end up with TypeError
+    // So, we need to set the prototype of attributes to null
+    var ObjectDefineProperty = function (obj, prop, attributes) {
+        _objectDefineProperty(obj, prop, setPrototype(attributes, null));
+    };
+    var CreateObject = platform.builtInJavascriptObjectCreate;
+
+    platform.registerChakraLibraryFunction("ArrayIterator", function (arrayObj, iterationKind) {
+        "use strict";
+        __chakraLibrary.InitInternalProperties(this, 4, "__$arrayObj$__", "__$nextIndex$__", "__$kind$__", "__$internalDone$__");
+        this.__$arrayObj$__ = arrayObj;
+        this.__$nextIndex$__ = 0;
+        this.__$kind$__ = iterationKind;
+        this.__$internalDone$__ = false; // We use this additional property to enable hoisting load of arrayObj outside the loop, we write to this property instead of the arrayObj
+    });
+
+    // ArrayIterator's prototype is the C++ Iterator, which is also the prototype for StringIterator, MapIterator etc
+    var iteratorPrototype = platform.GetIteratorPrototype();
+    // Establish prototype chain here
+    __chakraLibrary.ArrayIterator.prototype = CreateObject(iteratorPrototype);
+
+    ObjectDefineProperty(__chakraLibrary.ArrayIterator.prototype, 'next',
+        {
+            value: function () {
+                "use strict";
+                let o = this;
+
+                if (!(o instanceof __chakraLibrary.ArrayIterator)) {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                let a = o.__$arrayObj$__;
+                let value, done;
+
+                if (o.__$internalDone$__ === true) {
+                    value = undefined;
+                    done = true;
+                }
+                else {
+                    let index = o.__$nextIndex$__;
+                    let len = __chakraLibrary.isArray(a) ? a.length : __chakraLibrary.GetLength(a);
+
+                    if (index < len) { // < comparison should happen instead of >= , because len can be NaN
+                        let itemKind = o.__$kind$__;
+
+                        o.__$nextIndex$__ = index + 1;
+
+                        if (itemKind === 1 /*ArrayIterationKind.Value*/) {
+                            value = a[index];
+                        }
+                        else if (itemKind === 0 /*ArrayIterationKind.Key*/) { // TODO (megupta) : Use clean enums here ?
```

Comment:
At this point we probably need to write a JS coding guideline. FWIW the prevailing pattern in library JS is to do `} else if (...) {` on a single line (no newline between `}` and `else`).

Reply:
Done. I am squashing CR suggestions with my prior commit, please bear with it!